sdk/pinocchio/src/cpi.rs

```diff
@@ -184,11 +251,103 @@ pub unsafe fn invoke_signed_unchecked(
     instruction: &Instruction,
     accounts: &[Account],
     signers_seeds: &[Signer],
+) {
+    invoke_instruction_signed_unchecked(
+        instruction.program_id,
+        instruction.accounts,
+        instruction.data,
+        accounts,
+        signers_seeds,
+    );
+}
+
+/// Invoke a cross-program instruction.
+#[inline(always)]
+pub fn invoke_instruction<const ACCOUNTS: usize>(
+    program_id: &Pubkey,
+    priviledges: &[Priviledge; ACCOUNTS],
+    instruction_data: &[u8],
+    accounts: &[Account; ACCOUNTS],
+) -> ProgramResult {
+    invoke_instruction_signed(program_id, priviledges, instruction_data, accounts, &[])
+}
+
+/// Invoke a cross-program instruction with signatures.
+#[inline]
+pub fn invoke_instruction_signed<const ACCOUNTS: usize>(
+    program_id: &Pubkey,
+    priviledges: &[Priviledge; ACCOUNTS],
+    instruction_data: &[u8],
+    accounts: &[Account; ACCOUNTS],
+    signers_seeds: &[Signer],
+) -> ProgramResult {
+    let mut account_metas = MaybeUninit::<[AccountMeta; ACCOUNTS]>::uninit();
+    let mut account_metas_ptr = account_metas.as_mut_ptr() as *mut AccountMeta;
+
+    accounts
+        .iter()
+        .zip(priviledges.iter())
+        .try_for_each(|(account, priviledge)| {
+            // Check whether all account infos can be safely borrowed according
+            // to their mutability on the instruction or not.
+            if account.is_borrowed(match priviledge.0 {
+                true => BorrowState::Borrowed,
+                false => BorrowState::MutablyBorrowed,
+            }) {
+                return Err(ProgramError::AccountBorrowFailed);
+            }
+
+            // SAFETY: There are `ACCOUNTS` account metas.
+            unsafe {
+                account_metas_ptr.write(AccountMeta::new(
+                    account.key(),
+                    priviledge.0,
+                    priviledge.1,
+                ));
+
+                account_metas_ptr = account_metas_ptr.add(1);
+            }
+
+            Ok(())
+        })?;
```

Comment:
*bikeshedding*

You can also use iterators to avoid both an `unsafe` block and explicit pointer.
`MaybeUninit` has an example of how to [initialize an array element by element](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element).

```suggestion
    /*
     * use itertools::izip;
     */

    let mut account_metas = MaybeUninit::<[AccountMeta; ACCOUNTS]>::uninit();

    for (account, (is_writable, is_signer), account_meta) in
        izip!(accounts, priviledges, &mut account_meta)
    {
        // Check whether all account infos can be safely borrowed according
        // to their mutability on the instruction or not.
        if account.is_borrowed(if is_writable {
            BorrowState::Borrowed
        } else {
            BorrowState::MutablyBorrowed
        }) {
            return Err(ProgramError::AccountBorrowFailed);
        }

        account_meta.write(AccountMeta::new(account.key(), is_writable, is_signer));
    }
    
    // SAFETY: Previous loop only finishes when all elements have been initialized.
    let account_metas = unsafe { account_metas.assume_init_ref() };
```

Reply:
Nice, we can then remove the `enumerate` on the other iterations.